oak_runtime/introspection_browser_client/index.tsx

```diff
@@ -45,6 +45,104 @@ function loadSerializedEvents(): Promise<Uint8Array> {
   });
 }
 
+interface OakApplicationState {
+  nodeInfos: NodeInfos;
+  channels: Channels;
+}
+
+type NodeId = number;
+type AbiHandle = number;
+type NodeInfos = Map<NodeId, NodeInfo>;
+interface NodeInfo {
+  abiHandles: Map<AbiHandle, ChannelHalf>;
+}
+
+type ChannelID = number;
+enum ChannelHalfDirection {
+  Read,
+  Write,
+}
+interface ChannelHalf {
+  channelId: ChannelID;
+  direction: ChannelHalfDirection;
+}
+interface Message {
+  data: Uint8Array;
+  channels: ChannelHalf[];
+}
+interface Channel {
+  id: ChannelID;
+  messages: Message[];
+}
+type Channels = Map<ChannelID, Channel>;
+
+function eventReducer(
+  applicationState: OakApplicationState,
+  event: introspectionEventsProto.Event
+) {
+  const eventType = event.getEventDetailsCase();
+  const { EventDetailsCase } = introspectionEventsProto.Event;
+
+  switch (eventType) {
+    case EventDetailsCase.NODE_CREATED:
+      applicationState.nodeInfos.set(event.getNodeCreated().getNodeId(), {
+        abiHandles: new Map(),
+      });
+
+      break;
+    case EventDetailsCase.NODE_DESTROYED:
+      applicationState.nodeInfos.delete(event.getNodeDestroyed().getNodeId());
+
+      break;
+    case EventDetailsCase.CHANNEL_CREATED:
+      {
+        const channelId = event.getChannelCreated().getChannelId();
+        applicationState.channels.set(channelId, {
+          id: channelId,
+          messages: [],
+        });
+      }
+
+      break;
+    case EventDetailsCase.CHANNEL_DESTROYED:
+      applicationState.channels.delete(
+        event.getChannelDestroyed().getChannelId()
+      );
+
+      break;
+    case EventDetailsCase.HANDLE_CREATED:
+      {
+        const details = event.getHandleCreated();
+        applicationState.nodeInfos
+          .get(details.getNodeId())
+          .abiHandles.set(details.getHandle(), {
+            channelId: details.getChannelId(),
+            direction: 0,
+          });
+      }
+      break;
+    case EventDetailsCase.HANDLE_DESTROYED:
+      {
+        const details = event.getHandleDestroyed();
+        applicationState.nodeInfos
+          .get(details.getNodeId())
+          .abiHandles.delete(details.getHandle());
+      }
+      break;
+    case EventDetailsCase.MESSAGE_ENQUEUED:
+      // TODO(#913): Add support for displaying messages
+      break;
+    case EventDetailsCase.MESSAGE_DEQUEUED:
+      // TODO(#913): Add support for displaying messages
+      break;
+    default:
+      // This should never happen
+      throw new Error(`Encountered unhandled event of type ${eventType}`);
+  }
```

Comment:
This `switch` statement can basically be thought of as a poor man's version of pattern matching, implementing different logic for each event type. 

This would be a lot nicer if written Rust, but for now I felt that the overhead introduced by Rust/JS interop does not justify writing this logic in Rust, even if it would be nicer.